feat: add Date Selector component

manvinderjit opened the pull request:
## Summary

Add a Reusable Date Selector component for the design system

## Tests

Pending

## Additional information

Closes #31 

<img width="475" height="112" alt="Screenshot 2025-07-14 at 16-05-48 User Profile" src="https://github.com/user-attachments/assets/fa130c94-124f-4fc4-b672-08a8f074acd3" />

manvinderjit committed:
feat: add Date Selector component
add HTML for date selector component.
add css styles for date selector component.
add sample usage for date selector component.

kbventures commented on src/components/DateSelector/DateSelector.css:
```diff
@@ -0,0 +1,22 @@
+@layer components {
+	.dob-container {
```
.dob-container

Since I don't know what .dop is, it's safe to assume naming can be improved here?

madcampos replied:
Date of birth? 👀

manvinderjit replied:
I copied this from `CompleteProfile` where it was originally used for date of birth. But since it is reusable, i should just name it date.

kbventures commented on src/components/DateSelector/DateSelector.tsx:
```diff
@@ -0,0 +1,179 @@
+ * @param date - The date string to validate, expected in MM-DD format.
+ * @returns `true` if the format is valid, otherwise `false`.
+ */
+const isValidDateFormat = (date: string): boolean => {
```
Handling dates is notoriously prone to errors.  

Why are we not considering a library instead?

https://www.npmjs.com/package/date-fns

date-fns lets you import only the specific functions you need.

madcampos replied:
I'd rather use the [temporal polyfill](https://github.com/js-temporal/temporal-polyfill). But yes, handling dates is hard and error prone

manvinderjit replied:
@madcampos Should we use the temporal polyfill library or leave the code as is for now?

madcampos replied:
Leave it as is, we can revisit later

kbventures replied:
@manvinderjit @madcampos Another reason I mentioned considering a library: 29 will always be selectable—even in non-leap years. date-fns is highly modular, so we could use just what we need and ignore the rest. Create an issue with "Low priority" label.

madcampos replied:
https://github.com/torontojs/communityhub/issues/197

madcampos commented on src/components/DateSelector/DateSelector.tsx:
```diff
@@ -0,0 +1,179 @@
+
+interface Props {
+	labelContent?: string;
+	setDateValue: React.Dispatch<React.SetStateAction<string>>;
```
**issue**: changes to a component should be _events_ and not passing state setters down.

Please change it to a `onChange` event handler, and then the consumers for this component should set the state when the function is called.

manvinderjit replied:
I have changed it to a handleChange function because I only want to update the parent's state value when both day and month are selected and the date is valid. Not partially. 

More insights here:

https://github.com/torontojs/communityhub/pull/184#discussion_r2217361369

madcampos replied:
So, let's take the input element as an inspiration here. We can expose three pieces then:
- `onChange` event, when any of the things are changed. (`onInput` would be redundant here, think of this component similar to a checkbox that doesn't fire the `input` event.)
- The value of the component.
- The validation state of the component.

We still should expose the way to modify internal state to the component as "event handlers" instead of passing the setter function down.

madcampos commented on src/components/DateSelector/DateSelector.tsx:
```diff
@@ -0,0 +1,179 @@
+		});
+	};
+
+	useEffect(() => {
+		if (!isRealDate(tempDateValue)) {
+			setDateValue('');
+			return;
+		}
+		if (isRealDate(tempDateValue)) {
+			setDateValue(tempDateValue);
+		}
+	}, [tempDateValue]);
```
This `useEffect` is not needed. When the component prop changes it will get re-rendered anyways. We should instead use a function that receives the prop as input and does this validation.

_If, and only if_, the computation is expensive use a `useMemo` instead.

manvinderjit replied:
@madcampos I am using this useEffect for two reasons:

1. It only updates the `date` state of the parent component when date is valid, i.e., both Month and Day are selected. It keeps the state in the child component until the date is valid.
2. It resets the `date` state of the parent component when the date is invalid, e.g., a user goes from a Month like `March 31` to `Feb`. As switching to Feb won't have the 30 and 31 date, we need to reset it.

madcampos replied:
Thinking about the behaviour of elements like `input`, validation is lazy. We should do the same here and always update the value, but only validate when needed to.

Furthermore, the input itself should work to prevent users from inputting the wrong date.

manvinderjit replied:
@madcampos Since are using two different selects

<img width="475" height="112" alt="image" src="https://github.com/user-attachments/assets/93bc78a4-8882-4453-ab4b-5bf549ab529e" />

I want to prevent situations wherein user selects only either day or month, not both and then clicks submit

In such a case the state value for the date in parent component will be '09-' or '-12'. 

Should we just add this check at form submission then and not onChange?

madcampos replied:
We should add both: A message when the date is invalid, and then a check when submitting to block subissions on invalid date

manvinderjit committed:
fix: rename css class for clarity

manvinderjit committed:
fix: swap state setter with handle function

swap state setter with a handle change function defined in parent.
update the sample example usage for date component.

kbventures reviewed (dismissed):
Approved.  Great work .Please create low priority issue so we can deal with updating this  later.

manvinderjit committed:
Merge branch 'dev' into reusable-date-selector-component

manvinderjit committed:
Merge branch 'dev' into reusable-date-selector-component

manvinderjit committed:
fix: lift the date component states to parent, pass as props

lift the states for the DateComponent's value and error handling to parent.
add props and update types for props.
add error message to date component.

manvinderjit committed:
styles: fix styles for date component

manvinderjit committed:
fix: update the usage docs for date component

manvinderjit commented:
@madcampos Updated the component based on the feedback. Please review. Thanks.

bcha92 reviewed (dismissed):
Day selector appears to be fixed, only appearing the appropriate date (i.e. September should have 30 days, not 31).
Only need to merge the base branch.

madcampos commented on src/components/DateSelector/DateSelector.tsx:
```diff
@@ -1,9 +1,13 @@
 
 interface Props {
+	dateValue: string;
+	isDateValid: boolean | null;
```
**question**: Isn't this prop redundant in the sense that this can be derived here from the provided date and the validation being ran?

Like, the component already runs some validation and calls two handlers:
- When the value updates
- When the value is valid/invalid

manvinderjit replied:
@madcampos But the parent needs to know if the date is valid, to prevent form submission, whose state `isDateValid` will be in the parent.

madcampos replied:
Yes, the state will live on the parent, and is set by the validation handler. It does not need to be a prop on this component as it can be derived from the `dateValue`.

manvinderjit replied:
@madcampos Addressed, please review.

madcampos commented on src/components/DateSelector/DateSelector.tsx:
```diff
@@ -59,9 +63,8 @@ const getDaysInMonth = (month: string) => {
-	const [tempDateValue, setTempDateValue] = useState<string>('');
-	const [month, day] = tempDateValue ? tempDateValue.split('-') : ['', ''];
+const DateSelector = ({ dateValue, isDateValid, handleSetDateValue, handleSetDateValidity, labelContent = 'Select Date' }: Props) => {
+	const [month, day] = dateValue ? dateValue.split('-') : ['', ''];
```
```suggestion
	const [month = '', day = ''] = dateValue?.split('-') ?? [];
```

madcampos commented on src/components/DateSelector/DateSelector.tsx:
```diff
@@ -96,42 +99,36 @@ const DateSelector = ({ handleSetDateValue, labelContent = 'Select Date' }: Prop
-	}, [tempDateValue]);
+		if (dateValue.trim() === '') { return; }
+		handleSetDateValidity(isRealDate(dateValue));
+	}, [dateValue]);
```
This should be called after the date value change handler is called, not as an effect. It makes the code cleaner.

manvinderjit replied:
@madcampos How will this impact scenarios wherein the parent component clears all fields, like a reset button?

madcampos replied:
I think it won't be a problem as validation should only be triggered after user input, so the parent setting the value won't "count" as the user has not changed the input yet.

manvinderjit replied:
Addressed, please review.

manvinderjit committed:
fix: cleanup variable declaration for month, day

Co-authored-by: Marco Campos <me@madcampos.dev>

manvinderjit committed:
fix: remove isDateValid prop

remove the isDateValid prop and calculate it from date value.
update the design system example to conform to the change.

manvinderjit committed:
fix: check date validity in handler function

check date validity in handler function instead of a useEffect.

manvinderjit committed:
Merge changes from dev into reusable-date-selector-component

manvinderjit committed:
fix: resolve eslint errors

manvinderjit committed:
Merge branch 'dev' into reusable-date-selector-component

manvinderjit committed:
fix: linting errors

remove react import.
convert function property signature to method shorthand in ts interface.
use const for nonchanging variables.
fix code leading to no-numbers warnings for linting.
fix regex to resolve linting error.

manvinderjit committed:
fix: disable linting rule for code block

madcampos commented on src/components/DateSelector/DateSelector.tsx:
```diff
@@ -0,0 +1,189 @@
+ * This function uses a placeholder year (2000, a leap year) to account for month lengths,
+ * including February 29th. It assumes no year context is needed and is used primarily for validation.
+ *
+ * @deprecated We will remove update this when Temporal becomes stable
```
<!-- Praises highlight something positive. Try to leave at least one of these comments per review. Do not leave false praise (which can actually be damaging). Do look for something to sincerely praise. -->
**praise:** Nice!

manvinderjit replied:
Thank you for the valuable feedback that made it nice.
